File: Python/Number of Islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid:
            return 0
            
        rows, cols = len(grid), len(grid[0])
        count = 0
        stack = []
        
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == "0" or grid[i][j] == "2":
                    continue
                count += 1
                grid[i][j] = "2"
                stack.append((i, j))
                while stack:
                    r, c = stack.pop()
                    # look up
                    if r-1>=0 and grid[r-1][c]=="1":
                        stack.append((r-1, c))
                        grid[r-1][c] = "2"
                    # look right
                    if c+1<cols and grid[r][c+1]=="1":
                        stack.append((r, c+1))
                        grid[r][c+1] = "2"
                    # look down
                    if r+1<rows and grid[r+1][c]=="1":
                        stack.append((r+1, c))
                        grid[r+1][c] = "2"
                    # look left
                    if c-1>=0 and grid[r][c-1]=="1":
                        stack.append((r, c-1))
                        grid[r][c-1] = "2"
        return count
        
        

File: Java/Number of Islands.java
public class Solution {
    public int numIslands(char[][] grid) {
        if (grid==null || grid.length==0 || grid[0].length==0) 
            return 0;
            
        int rows = grid.length;
        int cols = grid[0].length;
        int count = 0;
        
        Stack<int[]> stack = new Stack<int[]>();
        
        for (int i=0; i<rows; i++) {
            for (int j=0; j<cols; j++) {
                if (grid[i][j]=='0' || grid[i][j]=='2') continue;
                count++;
                grid[i][j] = '2';
                int[] tmp = {i, j};
                stack.push(tmp);
                while (!stack.isEmpty()) {
                    int[] p = stack.pop();
                    int r = p[0];
                    int c = p[1];
                    
                    if (r-1>=0 && grid[r-1][c]=='1') {
                        grid[r-1][c] = '2';
                        int[] t = {r-1, c};
                        stack.push(t);
                    }
                    if (c-1>=0 && grid[r][c-1]=='1') {
                        grid[r][c-1] = '2';
                        int[] t = {r, c-1};
                        stack.push(t);
                    }
                    if (r+1<rows && grid[r+1][c]=='1') {
                        grid[r+1][c] = '2';
                        int[] t = {r+1, c};
                        stack.push(t);
                    }
                    if (c+1<cols && grid[r][c+1]=='1') {
                        grid[r][c+1] = '2';
                        int[] t = {r, c+1};
                        stack.push(t);
                    }    
                }
            }
        }
        
        return count;
    }
}